File: betamax-core/src/main/java/co/freeside/betamax/junit/RecorderRule.java
/*
 * Copyright 2013 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package co.freeside.betamax.junit;

import java.util.*;
import java.util.logging.*;
import co.freeside.betamax.*;
import org.junit.rules.*;
import org.junit.runner.*;
import org.junit.runners.model.*;
import static java.util.logging.Level.*;

/**
 * This is a wrapper for the Betamax {@link co.freeside.betamax.Recorder} that can be used as a
 * _JUnit @Rule_ allowing tests annotated with `@Betamax` to run with the
 * Betamax HTTP proxy in the background.
 */
public class RecorderRule implements TestRule {

    private final Recorder recorder;
    private final Logger log = Logger.getLogger(RecorderRule.class.getName());

    public RecorderRule(Recorder recorder) {
        this.recorder = recorder;
    }

    @Override
    public Statement apply(final Statement statement, Description description) {
        final Betamax annotation = description.getAnnotation(Betamax.class);
        if (annotation != null) {
            log.fine("found @Betamax annotation on '" + description.getDisplayName() + "'");
            return new Statement() {
                @Override
                public void evaluate() throws Throwable {
                    Map<String, Object> map = new LinkedHashMap<String, Object>(2);
                    map.put("mode", annotation.mode());
                    map.put("match", Arrays.asList(annotation.match()));
                    try {
                        recorder.start(annotation.tape(), map);
                        statement.evaluate();
                    } catch (Exception e) {
                        log.log(SEVERE, "Caught exception starting Betamax", e);
                    } finally {
                        recorder.stop();
                    }
                }
            };
        } else {
            log.fine(String.format("no @Betamax annotation on '%s'", description.getDisplayName()));
            return statement;
        }
    }

}

File: okreplay-junit/src/main/kotlin/okreplay/RecorderRule.kt
package okreplay

import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement
import java.util.logging.Level
import java.util.logging.Logger
import okreplay.CaseFormat.*

/**
 * This is an extension of [Recorder] that can be used as a
 * _JUnit @Rule_ allowing tests annotated with `@OkReplay` to automatically
 * activate OkReplay recording.
 */
class RecorderRule(configuration: OkReplayConfig) : Recorder(configuration), TestRule {
  override fun apply(statement: Statement, description: Description): Statement {
    val annotation = description.getAnnotation(OkReplay::class.java)
    if (annotation != null) {
      LOG.info(String.format("found @OkReplay annotation on '%s'", description.displayName))
      return object : Statement() {
        @Throws(Throwable::class)
        override fun evaluate() {
          try {
            val tapeName = if (annotation.tape.isNullOrEmpty()) {
              description.defaultTapeName()
            } else {
              annotation.tape
            }
            val tapeMode = annotation.mode
            val matchRules = annotation.match
            val matchRule = if (matchRules.isNotEmpty()) {
              Optional.of(ComposedMatchRule.of(*matchRules))
            } else {
              Optional.absent<MatchRule>()
            }
            start(tapeName, tapeMode.toOptional(), matchRule)
            statement.evaluate()
          } catch (e: Exception) {
            LOG.log(Level.SEVERE, "Caught exception starting OkReplay", e)
            throw e
          } finally {
            try {
              stop()
            } catch (e: IllegalStateException) {
              // Recorder has not started yet.
            }
          }
        }
      }
    } else {
      LOG.info(String.format("no @OkReplay annotation on '%s'", description.displayName))
      return statement
    }
  }

  private fun Description.defaultTapeName(): String {
    val name = if (methodName != null) {
      LOWER_CAMEL.to(LOWER_UNDERSCORE, methodName)
    } else {
      UPPER_CAMEL.to(LOWER_UNDERSCORE, testClass.simpleName)
    }
    return name.replace('_', ' ')
  }

  companion object {
    private val LOG = Logger.getLogger(RecorderRule::class.java.simpleName)
  }
}
